Add public maintenance process for security reports (#16537)
Qiskit/qiskit-terra: MAINTAINING.md

You will need to run `cargo build` as part of a version-bump commit to propagate the changes in `Cargo.toml` to `Cargo.lock`.

<span id="security"></span>
## Handling Security Vulnerabilities

Some bugs allow attackers to target systems using Qiskit as a library in ways that cannot reasonably be protected against, and can have severe negative consequences for those systems.
We call these "security vulnerabilities", and there are special processes for handling these.

Security vulnerabilities, including reports that _may_ be security vulnerabilities, are handled privately to avoid disclosing a potential attack surface until a fix or workaround is ready for deployment.
This poses several problems:

* even _potential_ security reports are high-priority interrupts on maintainers' time
* it is hard to collaborate on and validate fixes
* it is hard to deploy a fix without alerting bad actors to the presence of a vulnerability
* it is hard to train new maintainers on the processes

This section covers our maintenance policy for handling these.
Maintainers may also have further responsibilities to their employers, which are not documented here.

Further reading:

> * [Qiskit's security policy](/SECURITY.md) for what constitutes a security vulnerability and how to report one safely.

Reviewer: Does this imply a dependency on #16535 which adds the definition of what qiskit considers a vulnerability?

Author: Yeah, I guess.  I think I wrote this one second, at any rate.

* [GitHub's documentation on its security tooling](https://docs.github.com/code-security/how-tos/report-and-fix-vulnerabilities).

### General principles

* Vulnerability reports should be triaged quickly.

  We aim to provide an initial response to a report within one working day.

>   The reponse can be one of "accept", "reject" or "investigate further".

Reviewer 2: typo: reponse

Reviewer: ```suggestion
  The response can be one of "accept", "reject" or "investigate further".
```

  If we need to investigate further, we aim to produce a final response within a week.

> * If a report is accepted, a fix or workaround should be prepared and released as soon as possible for all affected Qiskit versions with active security support.

Reviewer: Do we have a thing to link to on this? Just so there is a reference to which branches this currently is?

Author: I think the best we've got is this: https://quantum.cloud.ibm.com/docs/en/guides/qiskit-sdk-version-strategy, though we could also update that page to have a dedicated section.

Author: I added that link in e8d8157c28c, and we can revisit if we like.

  We aim to release fixes within a month of accepting a report, and we must release the fix within three months of the report being made.

* There should be as little time as reasonably possible between public disclosure and the fix being available in a released version of Qiskit.

  This is a trade-off between the risk of vulnerability exploitation and the risk that rushed and insufficiently reviewed code is released.

* As few people as reasonably possible should be exposed to vulnerability reports.

>   This is a trade-off between the risk of accidental disclosure, the risk of burning out the administrators, and the need to train new maintainers on the security procedures.

Reviewer: Do we have an administrator role defined anywhere in this file. I think this might be the first time we use admin or administrator in this document so it might be worth being more explicit what you mean here.

Author: I might have shifted some lines around, because `admin` is (sort of) defined at the start of the next section - certainly all I mean right now is "people with `admin` permission on GitHub".

Author: I added a line into the intro section that deliberately uses `admin` to pre-define this in e8d8157c28c.

### Who is involved

Security reports are automatically visible to exactly those users with `admin` rights to the repository.
This is one reason that `admin` rights must be very limited.

There should be at least three people actively involved in any "accepted" or "investigate further" security report.

> One non-admin maintainer should be added to each such security report, in order to train others on the process.

Reviewer 2: I feel entire maintaining team should be able to see the report. Is there any concern here I am missing?

Reviewer 3: > This is a trade-off between the risk of accidental disclosure, the risk of burning out the administrators, and the need to train new maintainers on the security procedures.

The risk of accidental disclosure is why the full team does not see it.

> If there is only one active admin at a given time, two non-admin maintainers should be added intead.

Reviewer 2: typo: intead

When choosing the non-admin collaborator(s), consider that the total goals are:

* we need a diversity of opinions for patches that may be controversial;
* we need at least one person, and ideally two, with expertise in the affected code;
* we need to train new maintainers on the security procedures.

All security-related fixes must have **two** non-authoring approvers, using the standard repository
rules for when a contribution amounts to co-authorship and requires a different reviewer.

An "active" member is one who is significantly participating in the triage, remediation or review of a report or fix.

> Examples of "non-active" admins are organization-level admins whose purpose is not repository-specific, and those who are not currently working owing to illness or holiday.

Reviewer: Maybe make this a bit less specific to a corporate environment:

```suggestion
Examples of "non-active" admins are organization-level admins whose purpose is not repository-specific, and those who are not currently available (e.g. are on holiday).
```

(I'll let the holiday vs vacation, UK English vs US English slide :P)

Author: > I'll let the holiday vs vacation, UK English vs US English slide

damn it - I was trying hard with the spelling!

Author: Fixed in e8d8157c28c.

### Remediation process

Once accepted, a fix must be prepared and released, and the vulnerability must be publicly announced, in order to allow downstream users to update.

Employed administrators may have company-internal processes that need to happen in parallel to this
process.
For example, IBM employees have responsibilities to the internal [PSIRT processes](https://www.ibm.com/trust/security-vulnerability-management) that begin immediately, and should consult internal documentation on these.

**Remediation process**:

> * [Create a new private fork](https://docs.github.com/code-security/tutorials/fix-reported-vulnerabilities/collaborate-in-a-fork) for preparing the patches.

Reviewer: A minor formatting detail do we want this to be a numbered list with subpoints? Like I would expect this to be something like:

```
1. Prepare the private fix PR
  a. Create a private fork
  b. prepare the PR
  c. Security release note
  d. Create private PR off the security advisory draft
2. Review the PR draft
  ...
```

since there is an order of operations we need to follow.

Author: Done in e8d8157c28c.

  **Do not** push any work to Qiskit/qiskit or your own fork.

* Using the private fork, prepare a PR branched from `main` that fixes or invalidates the vulnerability.
  **Note**: there is no CI available on private forks.
  Take additional care to run the full test and lint suite locally, on all tier 1 platforms.

* Include a `security` release note in the PR.

* Have two collaborators review and approve the PR.
  After each round of changes, be sure to re-run the tests and lint.

> * Prepare cherry-pick PRs to each `stable/*` branch with security support.

Reviewer: I would mention specifically on the private fork here too. Just to make it clear for this step the goal is to have open private PRs for everything that will be published as part of the disclosure/fix process. This lets all parties involved in the vulnerability process review and sign off on everything privately first before we make it public. Then the disclosure process is about final validation/CI and publication as the review/approvals have already been done ahead of time.

Author: Done in e8d8157c28c.

  **Run the tests and lint again for each branch.**

> * For each `stable/*`-branch PR, follow the procedure for preparing a new patch release and add the version bumps to the PRs.

Reviewer: This doesn't call out the versioning here. It might be obvious following semver that this typically is just a new patch release for each `major.minor` that has security support. But we might want to call this out explicitly here.

Author: I said "prepar[e] a new patch release" - do you want the word "version" in there somewhere too?

Author: I stuck the word "version" in in e8d8157c28c.

When designing a fix, consider these points:

* Fixes should be as minimal as possible.

>   The process is more stressful and unverified than usual, and it is easy to introduce new bugs and vulnerabilities.

Reviewer: I would probably call out the risk of regression here too. Just like with stable backports that's what we're typically balancing here, and that's with the extra weight of the stress and abnormal process.

Author: Done in e8d8157c28c.

* Fixes do not need to be elegant.

  It is better to publish an overcautious fix immediately within the private process, then follow up with a cleaner solution under normal conditions.

>   For example, we fixed [a recursive stack overflow in `qiskit.qasm2.load`](https://github.com/Qiskit/qiskit/security/advisories/GHSA-w7g6-mx9c-q2hr) with [an ugly depth limit](https://github.com/Qiskit/qiskit/pull/16421) first, then replaced the parser with [a fully iterative one](https://github.com/Qiskit/qiskit/pull/16425) later.

Reviewer: Maybe or some other change to the language to explain why the depth limit was chosen rather than just calling it ugly?

```suggestion
  For example, we fixed [a recursive stack overflow in `qiskit.qasm2.load`](https://github.com/Qiskit/qiskit/security/advisories/GHSA-w7g6-mx9c-q2hr) with [an simple but not-ideal depth limit](https://github.com/Qiskit/qiskit/pull/16421) first, then replaced the parser with [a fully iterative one](https://github.com/Qiskit/qiskit/pull/16425) later.
```

Author: Done in e8d8157c28c.

> * Fixes take priority over backwards-compatibile stability.

Reviewer 2: typo: compatibile

Reviewer: ```suggestion
* Fixes take priority over backwards-compatible stability.
```

  Wherever possible, aim to avoid impact to legitimate uses of Qiskit, but do not let this compromise the publication of a fix.
  Consider adding limitations by optional keyword arguments with reasonable defaults, but can be explicitly lifted by users.

> 

Reviewer: Do you want to call out that security can take priority over normal backwards compatibility guarantees, performance considerations, etc here?

Author: That's on line 164, no?  I can add the bit about performance to that?

Author: Added in e8d8157c28c.

### Disclosure and release process

The order of disclosure and release for an accepted report is as follows:

1. Co-ordinate with the responsible CVE issuer (currently IBM PSIRT) to receive a complete CVSS score _specifically for Qiskit SDK_.

>    There may be additional CVEs issued against specific services using Qiskit SDK, which are not covered by this policy.

Reviewer: ```suggestion
   There may be additional CVEs issued against downstream users of Qiskit SDK, which are not covered by this policy.
```

Author: Done in e8d8157c28c.

   Fill in the advisory with this, and all other necessary details listed below this process.

2. Prepare and review the fixes for each supported branch, as in the above section.

3. Co-ordinate a disclosure date and time with the responsible owner (currently IBM PSIRT).

4. At the specified time, re-open the approved PRs, now on the *public* repository.
   **Do not publish the advisory yet.**
   Allow the complete CI suite to pass, after fixing any caught mistakes.

>    Once CI has passed, the PRs can be immediately admin-merged, skipping the merge queue.

Reviewer: We should mention here that the reviews required for the above section are why we can fast track this. The additional CI runs are required to ensure that we merge functional code and actively verify that before releasing it.

Author: Done in e8d8157c28c.

  For example, we fixed [a recursive stack overflow in `qiskit.qasm2.load`](https://github.com/Qiskit/qiskit/security/advisories/GHSA-w7g6-mx9c-q2hr) with [a simple but overly restrictive depth limit](https://github.com/Qiskit/qiskit/pull/16421) first, then replaced the parser with [a fully iterative one](https://github.com/Qiskit/qiskit/pull/16425) later.

* Fixes take priority over backwards-compatible stability and performance.

  Wherever possible, aim to avoid impact to legitimate uses of Qiskit, but do not let this compromise the publication of a fix.
  Consider adding limitations by optional keyword arguments with reasonable defaults, but can be explicitly lifted by users.

### Disclosure and release process

The order of disclosure and release for an accepted report is as follows:

1. Co-ordinate with the responsible CVE issuer (currently IBM PSIRT) to receive a complete CVSS score _specifically for Qiskit SDK_.
   There may be additional CVEs issued against downstream users of Qiskit SDK, which are not covered by this policy.

>    Fill in the advisory with this, and all other necessary details listed below this process.

Reviewer 4: Can you add a link to the definitions of `CVE`/`CVSS`?

Author: Done in 098a477be54.

When designing a fix, consider these points:

* Fixes should be as minimal as possible.

  The process is more stressful and unverified than usual, and it is easy to introduce new bugs and vulnerabilities, or regress on old ones.

* Fixes do not need to be elegant.

  It is better to publish an overcautious fix immediately within the private process, then follow up with a cleaner solution under normal conditions.
  For example, we fixed [a recursive stack overflow in `qiskit.qasm2.load`](https://github.com/Qiskit/qiskit/security/advisories/GHSA-w7g6-mx9c-q2hr) with [a simple but overly restrictive depth limit](https://github.com/Qiskit/qiskit/pull/16421) first, then replaced the parser with [a fully iterative one](https://github.com/Qiskit/qiskit/pull/16425) later.

* Fixes take priority over backwards-compatible stability and performance.

  Wherever possible, aim to avoid impact to legitimate uses of Qiskit, but do not let this compromise the publication of a fix.
  Consider adding limitations by optional keyword arguments with reasonable defaults, but can be explicitly lifted by users.

### Disclosure and release process

> The order of disclosure and release for an accepted report is as follows:

Reviewer: The thing we should maybe explicitly mention here is that the disclosure can only occur after we have approved the PRs on all supported branches in the private fork. It's implicit in the steps below, but just calling it out for clarity might be useful.

Author: I made step two "prepare, review and approve" instead of just "prepare and review" - is that clear enough?

Reviewer: Yeah I think that's enough

1. Co-ordinate with the responsible CVE[^1] issuer (currently IBM PSIRT) to receive a complete CVSS[^2] score _specifically for Qiskit SDK_.
   There may be additional CVEs issued against downstream users of Qiskit SDK, which are not covered by this policy.
   Fill in the advisory with this, and all other necessary details listed below this process.

2. Prepare and review the fixes for each supported branch, as in the above section.

3. Co-ordinate a disclosure date and time with the responsible owner (currently IBM PSIRT).

> 4. At the specified time, re-open the approved PRs, now on the *public* repository.

Reviewer: I think just a clarity point here would be to reword this as something like:

```suggestion
4. At the specified time, create a PR on the *public* repository from the previously approved PRs in the private fork.
```

Re-open might be a bit confusing here because it was never closed explicitly in the above steps.

Author: Done (similarish) in d470c09
